_posts/2025-02-05-#31363.md

```diff
@@ -10,17 +10,263 @@ status: upcoming
+- When a `TxGraph` is oversized, mutation functions (like `RemoveTransaction` and `AddDependency`) and
+  `GetIndividualFeerate` can still be called, but inspection functions like `` are unavailable because we won't
+linearize clusters that are oversized.
+- Additionally, step 4 will not proceed if the txgraph is oversized; step 5 is thus also impossible.
+
+The `GroupClusters` function implements [union-find](https://en.wikipedia.org/wiki/Disjoint-set_data_structure), a
+favorite amongst software engineering interviewers.
+
+<!-- FIXME: Unlinked -->
+
+#### Staging Support
+
+This section includes code up to ["txgraph: (feature) add staging support"](https://github.com/bitcoin-core-review-club/bitcoin/commit/79647fd58dc7152eb1dc4e09cc56db845608352d).
+
+This feature allows users of `TxGraph` to stage changes and then apply or discard them, much like git.
+
+Much of the `TxGraph` members are grouped together as a `ClusterSet`: ther vector of `Cluster`s, work queues, and
+counter. This refactor is done so that a `TxGraph` can have multiple **subgraphs** or **levels** represented as
+`ClusterSet`s: main (level 0) and staged (level 1).
+
+[Creating](https://github.com/bitcoin-core-review-club/bitcoin/commit/79647fd58dc7152eb1dc4e09cc56db845608352d#diff-c25ebcc6e4dc9c9bb21462f796fbb470cee3f0ff1e77bfc4033f4e8460f4482cR1464)
+a staged subgraph requires first applying removals. The staged subgraph creates copies of information from the main graph.
+
+<!-- FIXME: more on impl -->
+
+#### Specialized Functionality
+
+This section includes the rest of #31363, but also #31444.
+
+There are various functions added to `TxGraph` to obtain specialized information:
+- `CountDistinctClusters` and `GetMainStagingDiagram` are used in RBF logic.
+- `GetBlockBuilder` is used in block assembly.
+- `GetWorstMainChunk` is used in mempool eviction.
+
+It may be useful to look at the code surrounding mempool to understand why each of the functions are necessary.
+
+<!-- #### Discretional Size-Limiting -->
+<!-- This is everything up to Trim(), i.e. #31553 -->
 
 ## Questions
 
+<!-- Basic TxGraph -->
+
 1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)? What was your review approach?
--->
+
+1. What are the benefits of having this `TxGraph` within `CTxMemPool`, in your own words? Can you think of drawbacks?
+
+1. What is the difference between `TxGraph` and `Cluster`?
+
+1. If a txgraph is oversized, which functions are still guaranteed to return an accurate result, and which ones aren't?
+
+1. Under what circumstances could `GetIndividualFeerate` and `GetChunkFeerate` return nothing (i.e. an empty `FeeFrac`?)
+   Is it possible that one returns an answer and the other doesn't? How?
+
+1. Why is `AddTransaction` annotated with `[[nodiscard]]`?
+
+1. After calling `TxGraph::RemoveTransaction`, is the transaction removed transaction?
```

Comment:
I think there's a grammar issue here, perhaps you meant `transaction removed immediately`?